# ***Credit Risk Analysis: SQL-Driven Insights for Lending Strategy***

*This project analyzes 51,336 credit applicants to uncover behavioral patterns that distinguish high-risk from low-risk borrowers. Using SQL-driven feature engineering and cohort analysis, we debunk common myths about credit scoring and provide actionable recommendations for improving lending decisions.*

---



### **Let's Get Started: Setting Up Our Toolkit**

First things first - I need to make sure we have all the right tools for this credit analysis project. Just like a carpenter needs a hammer and nails, I need specific Python libraries to connect to our MySQL database and run SQL queries right from this Jupyter notebook.

In [ ]:
!pip install jupysql pymysql sqlalchemy

###  **Plugging Into Our Database**

Now I'm loading the SQL "magic" extension. This is what allows me to seamlessly mix Python and SQL in this notebook. It's like having a universal adapter that lets me speak both Python and SQL languages in the same conversation.

In [1]:
%load_ext sql

###  **Establishing Contact with Our Data Warehouse**

Time to connect to the heart of our project - the MySQL database. This is where all our credit data lives. I'm using my credentials to securely log into the server where we store information on thousands of loan applicants.

In [2]:

%sql mysql+pymysql://root:harsh@127.0.0.1:3306/db_p1

Connecting to 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

In [3]:
%%sql
SHOW TABLES;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

3 rows affected.

+------------------------+
|    Tables_in_db_p1     |
+------------------------+
| external_cibil_dataset |
|   final_credit_data    |
| internal_bank_dataset  |
+------------------------+

In [11]:
%%sql
SELECT *
FROM internal_bank_dataset AS IBD JOIN external_cibil_dataset AS ECD
ON IBD.PROSPECTID = ECD.PROSPECTID
LIMIT 10;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

10 rows affected.

+------------+----------+---------------+---------------+---------------------+-------------------+-----------------+-------------------+---------------+---------------+----------------------+--------------------+------------------+--------------------+-----------------+---------+-------+-------------+---------+---------+-------+------------+--------------+----------+---------------+---------------+--------------+---------------------------+-----------------------------+------------------------------+----------------------+-----------------------+---------------------------+----------------+-----------------+-------------------+----------------+-----------------+-------------------+-------------------+---------+--------------+---------------+---------+--------------+---------------+---------+--------------+---------------+---------+--------------+---------------+-----------------------+---------+--------+------------+-------------+--------+------------+-------------+-----------------------+----------+---------+---------+---------------+----------------+-----+--------+------------------+---------------------+------------------------+----------------------------+-----------------------+----------------+---------+----------------+---------+------------------------+------------------------+------------------------+------------------------+--------------------------+---------+---------+----------------+-----------------+--------------+---------------+
| PROSPECTID | Total_TL | Tot_Closed_TL | Tot_Active_TL | Total_TL_opened_L6M | Tot_TL_closed_L6M | pct_tl_open_L6M | pct_tl_closed_L6M | pct_active_tl | pct_closed_tl | Total_TL_opened_L12M | Tot_TL_closed_L12M | pct_tl_open_L12M | pct_tl_closed_L12M | Tot_Missed_Pmnt | Auto_TL | CC_TL | Consumer_TL | Gold_TL | Home_TL | PL_TL | Secured_TL | Unsecured_TL | Other_TL | Age_Oldest_TL | Age_Newest_TL | PROSPECTID_1 | time_since_recent_payment | time_since_first_deliquency | time_since_recent_deliquency | num_times_delinquent | max_delinquency_level | max_recent_level_of_deliq | num_deliq_6mts | num_deliq_12mts | num_deliq_6_12mts | max_deliq_6mts | max_deliq_12mts | num_times_30p_dpd | num_times_60p_dpd | num_std | num_std_6mts | num_std_12mts | num_sub | num_sub_6mts | num_sub_12mts | num_dbt | num_dbt_6mts | num_dbt_12mts | num_lss | num_lss_6mts | num_lss_12mts | recent_level_of_deliq | tot_enq | CC_enq | CC_enq_L6m | CC_enq_L12m | PL_enq | PL_enq_L6m | PL_enq_L12m | time_since_recent_enq | enq_L12m | enq_L6m | enq_L3m | MARITALSTATUS |   EDUCATION    | AGE | GENDER | NETMONTHLYINCOME | Time_With_Curr_Empr | pct_of_active_TLs_ever | pct_opened_TLs_L6m_of_L12m | pct_currentBal_all_TL | CC_utilization | CC_Flag | PL_utilization | PL_Flag | pct_PL_enq_L6m_of_L12m | pct_CC_enq_L6m_of_L12m | pct_PL_enq_L6m_of_ever | pct_CC_enq_L6m_of_ever | max_unsec_exposure_inPct | HL_Flag | GL_Flag | last_prod_enq2 | first_prod_enq2 | Credit_Score | Approved_Flag |
+------------+----------+---------------+---------------+---------------------+-------------------+-----------------+-------------------+---------------+---------------+----------------------+--------------------+------------------+--------------------+-----------------+---------+-------+-------------+---------+---------+-------+------------+--------------+----------+---------------+---------------+--------------+---------------------------+-----------------------------+------------------------------+----------------------+-----------------------+---------------------------+----------------+-----------------+-------------------+----------------+-----------------+-------------------+-------------------+---------+--------------+---------------+---------+--------------+---------------+---------+--------------+---------------+---------+--------------+---------------+-----------------------+---------+--------+------------+-------------+--------+------------+-------------+-----------------------+----------+---------+---------+---------------+------

### **Let's peek inside the data before we get fancy**

When I first connected to the database, I wanted to see what we actually had to work with. Two raw tables stared back at me: one with what the bank knows about applicants (income, job, age), and another packed with their credit bureau history (payments, loans, missed dues).

But here's the thing — 88 columns of data is overwhelming. Most of it was noise: duplicate IDs, mathematically redundant ratios (`closed = 1 - active`), and stale history that wouldn't help predict tomorrow's behavior.

So I made a call: keep only what matters *right now*. Recent behavior (last 3-6 months) for spotting desperation. Longer patterns (12 months) for judging character. And hard numbers — not ratios of ratios that break when denominators hit zero.

The result? A clean 47-column view where every feature has a plain-English reason to exist.

In [5]:
%%sql
CREATE OR REPLACE VIEW final_credit_data AS
SELECT 
    IBD.PROSPECTID,
    -- Wrapping numeric columns in NULLIF to handle -99999
    NULLIF(Credit_Score, -99999) AS Credit_Score,
    Approved_Flag,
    
    NULLIF(AGE, -99999) AS AGE,
    GENDER, MARITALSTATUS, EDUCATION, 
    NULLIF(NETMONTHLYINCOME, -99999) AS NETMONTHLYINCOME, 
    NULLIF(Time_With_Curr_Empr, -99999) AS Time_With_Curr_Empr,
    
    NULLIF(Total_TL, -99999) AS Total_TL,
    NULLIF(Tot_Active_TL, -99999) AS Tot_Active_TL,
    NULLIF(pct_active_tl, -99999) AS pct_active_tl, 
    NULLIF( pct_closed_tl, -99999) AS pct_closed_tl,
    NULLIF(Auto_TL, -99999) AS Auto_TL,
    NULLIF(CC_TL, -99999) AS CC_TL,
    NULLIF(Consumer_TL, -99999) AS Consumer_TL,
    NULLIF(Gold_TL, -99999) AS Gold_TL,
    NULLIF(Home_TL, -99999) AS Home_TL,
    NULLIF(PL_TL, -99999) AS PL_TL,
    NULLIF(Secured_TL, -99999) AS Secured_TL,
    NULLIF(Unsecured_TL, -99999) AS Unsecured_TL,
    NULLIF(Age_Oldest_TL, -99999) AS Age_Oldest_TL,
    NULLIF(Age_Newest_TL, -99999) AS Age_Newest_TL,

    NULLIF(Total_TL_opened_L6M, -99999) AS Total_TL_opened_L6M,
    NULLIF(pct_tl_open_L6M, -99999) AS pct_tl_open_L6M,
    NULLIF(Tot_TL_closed_L6M, -99999) AS Tot_TL_closed_L6M,
    NULLIF(enq_L6m, -99999) AS enq_L6m,
    NULLIF(enq_L3m, -99999) AS enq_L3m,
    NULLIF(time_since_recent_enq, -99999) AS time_since_recent_enq,

    NULLIF(num_deliq_12mts, -99999) AS num_deliq_12mts,
    NULLIF(max_deliq_12mts, -99999) AS max_deliq_12mts,
    NULLIF(num_std_12mts, -99999) AS num_std_12mts,
    NULLIF(num_sub_12mts, -99999) AS num_sub_12mts,
    NULLIF(num_dbt_12mts, -99999) AS num_dbt_12mts,
    NULLIF(num_lss_12mts, -99999) AS num_lss_12mts,

    NULLIF(Tot_Missed_Pmnt, -99999) AS Tot_Missed_Pmnt,
    NULLIF(num_times_delinquent, -99999) AS num_times_delinquent,
    NULLIF(time_since_recent_deliquency, -99999) AS time_since_recent_deliquency,
    NULLIF(time_since_recent_payment, -99999) AS time_since_recent_payment,
    NULLIF(num_times_30p_dpd, -99999) AS num_times_30p_dpd,
    NULLIF(num_times_60p_dpd, -99999) AS num_times_60p_dpd,
    NULLIF(max_delinquency_level, -99999) AS max_delinquency_level,
    NULLIF(recent_level_of_deliq, -99999) AS recent_level_of_deliq,

    NULLIF(CC_utilization, -99999) AS CC_utilization,
    NULLIF(PL_utilization, -99999) AS PL_utilization,
    NULLIF(max_unsec_exposure_inPct, -99999) AS max_unsec_exposure_inPct

FROM internal_bank_dataset AS IBD
JOIN external_cibil_dataset AS ECD ON IBD.PROSPECTID = ECD.PROSPECTID;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

++
||
++
++

In [7]:
%sql SELECT * FROM final_credit_data;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

51336 rows affected.

+------------+--------------+---------------+-----+--------+---------------+----------------+------------------+---------------------+----------+---------------+---------------+---------------+---------+-------+-------------+---------+---------+-------+------------+--------------+---------------+---------------+---------------------+-----------------+-------------------+---------+---------+-----------------------+-----------------+-----------------+---------------+---------------+---------------+---------------+-----------------+----------------------+------------------------------+---------------------------+-------------------+-------------------+-----------------------+-----------------------+----------------+----------------+--------------------------+
| PROSPECTID | Credit_Score | Approved_Flag | AGE | GENDER | MARITALSTATUS |   EDUCATION    | NETMONTHLYINCOME | Time_With_Curr_Empr | Total_TL | Tot_Active_TL | pct_active_tl | pct_closed_tl | Auto_TL | CC_TL | Consumer_TL | Gold_TL | Home_TL | PL_TL | Secured_TL | Unsecured_TL | Age_Oldest_TL | Age_Newest_TL | Total_TL_opened_L6M | pct_tl_open_L6M | Tot_TL_closed_L6M | enq_L6m | enq_L3m | time_since_recent_enq | num_deliq_12mts | max_deliq_12mts | num_std_12mts | num_sub_12mts | num_dbt_12mts | num_lss_12mts | Tot_Missed_Pmnt | num_times_delinquent | time_since_recent_deliquency | time_since_recent_payment | num_times_30p_dpd | num_times_60p_dpd | max_delinquency_level | recent_level_of_deliq | CC_utilization | PL_utilization | max_unsec_exposure_inPct |
+------------+--------------+---------------+-----+--------+---------------+----------------+------------------+---------------------+----------+---------------+---------------+---------------+---------+-------+-------------+---------+---------+-------+------------+--------------+---------------+---------------+---------------------+-----------------+-------------------+---------+---------+-----------------------+-----------------+-----------------+---------------+---------------+---------------+---------------+-----------------+----------------------+------------------------------+---------------------------+-------------------+-------------------+-----------------------+-----------------------+----------------+----------------+--------------------------+
|     1      |     696      |       P2      |  48 |   M    |    Married    |      12TH      |      51000       |         114         |    5     |       1       |      0.2      |      0.8      |    0    |   0   |      0      |    1    |    0    |   4   |     1      |      4       |       72      |       18      |          0          |       0.0       |         0         |    0    |    0    |          566          |        0        |       None      |       11      |       0       |       0       |       0       |        0        |          11          |              15              |            549            |         0         |         0         |           29          |           29          |      None      |     0.798      |          13.333          |
|     2      |     685      |       P2      |  23 |   F    |     Single    |    GRADUATE    |      19000       |          50         |    1     |       1       |      1.0      |      0.0      |    0    |   0   |      1      |    0    |    0    |   0   |     0      |      1       |       7       |       7       |          0          |       0.0       |         0         |    0    |    0    |          209          |        0        |        0        |       0       |       0       |       0       |       0       |        0        |          0           |             None             |             47            |         0         |         0         |          None         |           0           |      None      |      None      |           0.86           |
|     3      |     693      |       P2      |  40 |   M    |    Married    |      SSC       |        18        |         191         |    8     |       8       |    

## **Feature Selection & Dimensionality Reduction Summary**

The dataset was reduced from **88 → 47 features** using **parsimony** and **banking domain logic**, with a strong emphasis on **recent borrower behavior over historical noise**.

### Key Feature Removal Logic

- **Redundant IDs**  
  Duplicate primary keys with no variance or predictive value were removed.

- **Multicollinearity**  
  Mathematically dependent features were dropped to avoid perfect correlation and ensure model stability.

- **Temporal Optimization**  
  - **6-month window:** Captures credit hunger and short-term velocity  
  - **12-month window:** Captures payment behavior and stability  
  Overlapping temporal features were removed to preserve signal clarity.

- **Categorical Flag Redundancy**  
  Binary loan flags were removed since numerical loan counts already encode both presence and intensity.

- **Recency over History**  
  Recent delinquency metrics were retained; first or very old delinquency indicators were dropped.

- **Noisy Derived Ratios**  
  Complex “ratio of ratios” features were removed due to high variance and outlier sensitivity.

- **Lifetime Aggregates**  
  Lifetime totals were excluded in favor of recent (6–12 month) behavior, which is more predictive for credit risk.

- **Low Impact / High Cardinality Features**  
  Low-variance metrics and high-cardinality string features were removed.


### Final Feature View (47 Features)

| Dimension | Focus |
|---------|------|
| Identity & Target | IDs, score, default flag |
| Demographics | Age, income, education |
| Portfolio Mix | Loan counts & types |
| Credit History Age | Oldest & newest trade lines |
| Velocity (6M) | Recent openings & enquiries |
| Stability (12M) | Delinquency & account status |
| Risk Signals | Missed payments & severity |
| Stress Metrics | CC/PL utilization & exposure |

---
---



### **The Big Picture Question**

Before we get into detailed analysis, I want to understand the basic composition of this dataset. What percentage of applicants fall into each risk category? This sets the stage for everything that follows.

In [11]:
%%sql
SELECT 
SUM(CASE WHEN Approved_Flag = 'P1' THEN 1 ELSE 0 END)/(COUNT(*)) * 100 AS pct_cat_1,
SUM(CASE WHEN Approved_Flag = 'P2' THEN 1 ELSE 0 END)/(COUNT(*)) * 100 AS pct_cat_2,
SUM(CASE WHEN Approved_Flag = 'P3' THEN 1 ELSE 0 END)/(COUNT(*)) * 100 AS pct_cat_3,
SUM(CASE WHEN Approved_Flag = 'P4' THEN 1 ELSE 0 END)/(COUNT(*)) * 100 AS pct_cat_4
FROM final_credit_data;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

1 rows affected.

+-----------+-----------+-----------+-----------+
| pct_cat_1 | pct_cat_2 | pct_cat_3 | pct_cat_4 |
+-----------+-----------+-----------+-----------+
|  11.3040  |  62.7221  |  14.5161  |  11.4578  |
+-----------+-----------+-----------+-----------+

### **The Risk Breakdown - A Surprising Spread**

Here's what I found: the dataset is dominated by P2 applicants - that's the "good but not perfect" category at about 63%. That's actually healthy for a lending portfolio.

What's interesting is that the "elite" P1 group is only about 11%, which makes sense because truly excellent credit is rare.

But here's what caught my attention: about 26% of applicants (P3 + P4) are in the higher-risk categories. That's 1 in 4 people who might struggle to repay if the economy slows down.

This distribution tells me we're looking at a typical lending portfolio - mostly good customers, with a solid chunk of higher-risk ones that need careful handling.

---

### **Testing a Common Assumption**

Most people assume that more education means better credit management. Let me check if that's actually true in this dataset by first seeing what education categories exist.

In [12]:
%%sql
SELECT DISTINCT EDUCATION
FROM final_credit_data;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

7 rows affected.

+----------------+
|   EDUCATION    |
+----------------+
|      12TH      |
|    GRADUATE    |
|      SSC       |
| POST-GRADUATE  |
| UNDER GRADUATE |
|     OTHERS     |
|  PROFESSIONAL  |
+----------------+

### **The Education Test**

Let me check if college-educated applicants are actually less risky. I'm defining "educated" as having at least a graduate degree and seeing what percentage they represent in each risk tier.

In [13]:
%%sql
SELECT 
Approved_Flag,
COUNT(*) AS total_prospects,
SUM(CASE WHEN EDUCATION IN ('GRADUATE', 'POST-GRADUATE', 'UNDER GRADUATE', 'PROFESSIONAL') THEN 1 ELSE 0 END) AS educated_count,
ROUND(SUM(CASE WHEN EDUCATION IN ('GRADUATE', 'POST-GRADUATE', 'UNDER GRADUATE', 'PROFESSIONAL') THEN 1 ELSE 0 END) / COUNT(*) * 100, 2) AS pct_educated
FROM final_credit_data
GROUP BY 1
ORDER BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+-----------------+----------------+--------------+
| Approved_Flag | total_prospects | educated_count | pct_educated |
+---------------+-----------------+----------------+--------------+
|       P1      |       5803      |      3010      |    51.87     |
|       P2      |      32199      |     14960      |    46.46     |
|       P3      |       7452      |      3572      |    47.93     |
|       P4      |       5882      |      3134      |    53.28     |
+---------------+-----------------+----------------+--------------+

### **The Education Paradox Revealed**

This is fascinating - and counterintuitive!

I found that the HIGHEST percentage of educated people is actually in the P4 (highest risk) category at 53%! Meanwhile, the elite P1 group has about 52% educated.

That tiny difference tells a huge story: education level barely distinguishes between our best and worst credit risks.

What this suggests is that while education might help people earn more, it doesn't necessarily teach financial discipline. Some of the riskiest applicants might be high-earning professionals who are over-leveraged or managing their money poorly.

---

### **Testing Another Life Factor**

Since education wasn't a great predictor, let me check marital status. Common wisdom says married people are more stable - let's see if the data agrees.

In [14]:
%%sql
SELECT Approved_Flag,
COUNT(*) AS total_prospects,
SUM(CASE WHEN MARITALSTATUS = 'MARRIED' THEN 1 ELSE 0 END) AS married_count,
ROUND(SUM(CASE WHEN MARITALSTATUS = 'MARRIED' THEN 1 ELSE 0 END) / COUNT(*) * 100, 2) AS pct_married
FROM final_credit_data
GROUP BY 1
ORDER BY 1;


Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+-----------------+---------------+-------------+
| Approved_Flag | total_prospects | married_count | pct_married |
+---------------+-----------------+---------------+-------------+
|       P1      |       5803      |      5215     |    89.87    |
|       P2      |      32199      |     23697     |    73.60    |
|       P3      |       7452      |      5087     |    68.26    |
|       P4      |       5882      |      3753     |    63.80    |
+---------------+-----------------+---------------+-------------+

### **Marriage as a Stability Signal**

Now THIS is interesting! I'm seeing a clear pattern:

The P1 (elite) group is almost 90% married, while as we move down to P4, that percentage drops to about 64%.

That's a strong correlation! It makes intuitive sense too - married people often have more stability, shared financial responsibilities, and maybe more motivation to maintain good credit (buying homes, planning for kids' education).

This tells me that in credit scoring, marital status might be a much more useful signal than education level.

---

### **The Income Question**

Let me check if higher income automatically means better credit. I'll split applicants into income quartiles and see how they distribute across risk categories.

In [ ]:
%%sql 
WITH income_quartiles AS (
    SELECT PROSPECTID, NETMONTHLYINCOME, NTILE(4) OVER (ORDER BY NETMONTHLYINCOME) AS income_quartile
    FROM final_credit_data
)
SELECT FCD.Approved_Flag,
ROUND(SUM(CASE WHEN IQ.income_quartile = 4 THEN 1 ELSE 0 END)/COUNT(*) * 100, 2) AS high_income_percentage,
ROUND(SUM(CASE WHEN IQ.income_quartile IN (2,3) THEN 1 ELSE 0 END)/COUNT(*) * 100, 2) AS moderate_income_percentage,
ROUND(SUM(CASE WHEN IQ.income_quartile = 1 THEN 1 ELSE 0 END)/COUNT(*) * 100, 2) AS low_income_percentage
FROM final_credit_data FCD JOIN income_quartiles IQ ON FCD.PROSPECTID = IQ.PROSPECTID
GROUP BY FCD.Approved_Flag
ORDER BY FCD.Approved_Flag;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+------------------------+----------------------------+-----------------------+
| Approved_Flag | high_income_percentage | moderate_income_percentage | low_income_percentage |
+---------------+------------------------+----------------------------+-----------------------+
|       P1      |         32.64          |           49.11            |         18.25         |
|       P2      |         23.78          |           49.60            |         26.62         |
|       P3      |         23.36          |           50.70            |         25.94         |
|       P4      |         26.22          |           52.18            |         21.61         |
+---------------+------------------------+----------------------------+-----------------------+

### **Capacity vs. Character - A Clear Distinction**

The results here tell an important story about financial behavior:

Yes, the P1 (elite) group has the highest percentage of top earners (about 33%), which makes sense - more income means easier debt repayment.

But here's the interesting part: the P4 (highest risk) group actually has MORE high-income earners than the P2 and P3 groups!

This reveals a crucial insight: having the **capacity** to pay (high income) doesn't guarantee the **character** to pay. Some high earners are clearly struggling with debt management despite their salaries.

Meanwhile, about 18% of our elite P1 borrowers are actually low-income. This proves that financial discipline matters more than paycheck size.

---

### **The "Credit Hunger" Signal**

Now let's look at something more behavioral - credit enquiries. When people apply for multiple loans in a short period, it often signals financial stress. Let me check if high enquiry rates correlate with higher risk.

In [ ]:
%%sql
WITH deciles AS (
    SELECT PROSPECTID, enq_L3m, NTILE(10) OVER (ORDER BY enq_L3m) AS enq_L3m_decile, enq_L6m, NTILE(10) OVER (ORDER BY enq_L6m) AS enq_L6m_decile
    FROM final_credit_data
)
SELECT f.Approved_Flag,
COUNT(*) AS total_prospects,
SUM(CASE WHEN d.enq_L3m_decile >= 8 THEN 1 ELSE 0 END) AS high_enq_L3m_count,
SUM(CASE WHEN d.enq_L6m_decile >= 8 THEN 1 ELSE 0 END) AS high_enq_L6m_count,
ROUND(SUM(CASE WHEN d.enq_L3m_decile >= 8 THEN 1 ELSE 0 END)/COUNT(*) * 100,2) AS high_enq_L3m_pct,
ROUND(SUM(CASE WHEN d.enq_L6m_decile >= 8 THEN 1 ELSE 0 END)/COUNT(*) * 100,2) AS high_enq_L6m_pct
FROM deciles d
JOIN final_credit_data f ON d.PROSPECTID = f.PROSPECTID
GROUP BY f.Approved_Flag
ORDER BY f.Approved_Flag;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+-----------------+--------------------+--------------------+------------------+------------------+
| Approved_Flag | total_prospects | high_enq_L3m_count | high_enq_L6m_count | high_enq_L3m_pct | high_enq_L6m_pct |
+---------------+-----------------+--------------------+--------------------+------------------+------------------+
|       P1      |       5803      |        692         |        720         |      11.92       |      12.41       |
|       P2      |      32199      |        6126        |        6438        |      19.03       |      19.99       |
|       P3      |       7452      |        3746        |        3590        |      50.27       |      48.17       |
|       P4      |       5882      |        4835        |        4651        |      82.20       |      79.07       |
+---------------+-----------------+--------------------+--------------------+------------------+------------------+

### **Enquiry Patterns Tell a Clear Story**

The pattern here is striking and clear:

Looking at recent enquiries (last 3-6 months), I can see that as risk increases, enquiry rates skyrocket. 

The P4 (highest risk) group has over 80% of people in the top enquiry brackets! Meanwhile, the P1 (elite) group has only about 12%.

This makes perfect sense - people in financial trouble often apply for multiple loans, creating a "desperation loop" where each rejection makes them apply for more, which lowers their score further.

What's telling is that the jump from P2 (~19% high enquiries) to P3 (~50%) is massive. That's a clear threshold where credit-seeking behavior becomes a major red flag.

---

### **The "Maxed Out" Problem**

Credit card utilization (how much of your limit you're using) is another key behavioral signal. Let me check what percentage of cardholders in each risk category are using more than 70% of their limits.

In [15]:
%%sql
SELECT Approved_Flag,
(ROUND(SUM(CASE WHEN CC_utilization <= 0.3 THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2)) AS safe_cc_utilization,
(ROUND(SUM(CASE WHEN CC_utilization BETWEEN 0.3 AND 0.7 THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2)) AS moderate_cc_utilization,
(ROUND(SUM(CASE WHEN CC_utilization > 0.7 THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2)) AS high_cc_utilization
FROM final_credit_data
WHERE CC_utilization IS NOT NULL
GROUP BY 1
ORDER BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+---------------------+-------------------------+---------------------+
| Approved_Flag | safe_cc_utilization | moderate_cc_utilization | high_cc_utilization |
+---------------+---------------------+-------------------------+---------------------+
|       P1      |        40.86        |          20.07          |        39.07        |
|       P2      |        29.66        |          19.56          |        50.93        |
|       P3      |        24.78        |          15.94          |        59.27        |
|       P4      |        19.96        |          15.55          |        64.66        |
+---------------+---------------------+-------------------------+---------------------+

### **Utilization Reveals Financial Stress**

This analysis shows another clear pattern: financial stress increases with risk category.

In the P4 group, nearly 65% of credit card holders are using more than 70% of their limits! That's 2 out of 3 high-risk borrowers essentially maxed out on their cards.

Compare that to the P1 group, where only about 39% have high utilization.

What's interesting is the difference in *why* they might have high utilization. For P1 borrowers, it could be "spend a lot, pay it off every month" behavior. For P4 borrowers, it's likely "living on credit cards" because they're cash-strapped.

This utilization signal is powerful because it shows current financial stress, not just past behavior.

---

### **The "Skin in the Game" Factor**

Now let me check what percentage of each borrower's total debt is unsecured (no collateral like a house or car). Unsecured debt is riskier for lenders because there's nothing to repossess if the borrower defaults.

In [16]:
%%sql
SELECT Approved_Flag,AVG(Unsecured_TL/Total_TL)*100 AS avg_pct_unsecured_TL
FROM final_credit_data
WHERE Total_TL IS NOT NULL AND Total_TL != 0
GROUP BY 1
ORDER BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+----------------------+
| Approved_Flag | avg_pct_unsecured_TL |
+---------------+----------------------+
|       P1      |     34.06655860      |
|       P2      |     44.30609580      |
|       P3      |     54.25703430      |
|       P4      |     59.16872150      |
+---------------+----------------------+

### **Clear Progression in Unsecured Debt**

The pattern here couldn't be clearer: as risk increases, so does the percentage of unsecured debt.

P1 borrowers have only about 34% of their debt as unsecured, while P4 borrowers have nearly 60%!

This tells an important story about financial behavior: safer borrowers tend to have more "skin in the game" - they're borrowing against assets like homes or cars. Riskier borrowers are relying more on signature loans and credit cards with no collateral.

From a bank's perspective, this matters because unsecured debt has much lower recovery rates in default. A borrower with 60% unsecured debt represents significantly more potential loss.

---

### **Past Payment Problems - The Full Story**

I want to look at payment behavior from multiple angles. Let me check three things:
1. What percentage of people have NEVER missed a payment?
2. How many have had serious payment issues (more than 3 missed payments)?
3. What's the worst delinquency they've ever had?

In [17]:
%%sql
SELECT Approved_Flag,
ROUND(SUM(CASE WHEN num_times_delinquent = 0 OR num_times_delinquent IS NULL THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_no_delinquency,
ROUND(SUM(CASE WHEN Tot_Missed_Pmnt > 3 THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_high_missed_payments,
AVG(max_delinquency_level) AS avg_max_delinquency_level
FROM final_credit_data
GROUP BY 1
ORDER BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+--------------------+--------------------------+---------------------------+
| Approved_Flag | pct_no_delinquency | pct_high_missed_payments | avg_max_delinquency_level |
+---------------+--------------------+--------------------------+---------------------------+
|       P1      |       75.12        |           6.12           |          97.2812          |
|       P2      |       76.74        |           1.37           |          69.7758          |
|       P3      |       55.84        |           1.56           |          63.1054          |
|       P4      |       46.21        |           2.43           |          84.9352          |
+---------------+--------------------+--------------------------+---------------------------+

### **The Delinquency Story Has Two Parts**

This analysis revealed something interesting that needs two explanations:

First, the percentage of people with NO delinquency history is actually highest in P2 (77%), not P1 (75%). That seems counterintuitive at first.

But here's the key insight from the average max delinquency level: P1 borrowers actually have the WORST historical delinquency (97 days on average)!

This tells me that the P1 group isn't made of people with perfect histories. Instead, they're people who may have had serious problems in the distant past but have recovered and maintained clean records recently.

Meanwhile, P4 borrowers show both recent problems (46% have no delinquency sounds wrong - let me check... actually 46% have NO delinquency? That seems odd. Wait, I need to look at recent delinquency separately).

---

### **Recent vs. Historical Problems**

Let me dig deeper into delinquency by looking specifically at 30-day and 60-day late payments. These are more standardized measures that might tell a clearer story.

In [18]:
%%sql
SELECT Approved_Flag,
AVG(num_times_30p_dpd) AS avg_num_times_30p_dpd,
AVG(num_times_60p_dpd) AS avg_num_times_60p_dpd,
ROUND(SUM(CASE WHEN num_times_30p_dpd = 0 OR num_times_30p_dpd IS NULL THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_no_30p_delinquency,
ROUND(SUM(CASE WHEN num_times_60p_dpd = 0 OR num_times_60p_dpd IS NULL THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_no_60p_delinquency
FROM final_credit_data
GROUP BY 1
ORDER BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+-----------------------+-----------------------+------------------------+------------------------+
| Approved_Flag | avg_num_times_30p_dpd | avg_num_times_60p_dpd | pct_no_30p_delinquency | pct_no_60p_delinquency |
+---------------+-----------------------+-----------------------+------------------------+------------------------+
|       P1      |         0.8463        |         0.5182        |         84.96          |         89.80          |
|       P2      |         0.5480        |         0.3063        |         88.04          |         92.75          |
|       P3      |         0.9239        |         0.5204        |         78.62          |         87.56          |
|       P4      |         1.3560        |         0.8795        |         71.52          |         82.42          |
+---------------+-----------------------+-----------------------+------------------------+------------------------+

### **Recent Delinquency Tells the True Story**

Now the pattern becomes clearer! Looking at more recent and standardized delinquency measures:

P4 borrowers have the highest rates of both 30-day and 60-day late payments. About 28% of P4 borrowers have had a 30-day late payment, versus only 15% for P1.

What's really telling is the 60-day late payments - these are more serious. Only about 10% of P1 borrowers have had these, but nearly 18% of P4 borrowers have.

This confirms that while P1 borrowers might have old scars, they're maintaining clean recent records. P4 borrowers are showing current, ongoing payment problems.

---

### **The Most Important Time Window**

Let me focus on the most recent 12 months. In credit analysis, recent behavior matters more than ancient history. Let me check delinquency specifically in the last year.

In [19]:
%%sql
SELECT Approved_Flag,
AVG(time_since_recent_deliquency) AS avg_time_since_recent_delinquency,
AVG(num_deliq_12mts) AS avg_num_times_12m_delinquency,
ROUND(SUM(CASE WHEN num_deliq_12mts = 0 OR num_deliq_12mts IS NULL THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_no_12m_delinquency
FROM final_credit_data
GROUP BY 1
ORDER BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------+-----------------------------------+-------------------------------+------------------------+
| Approved_Flag | avg_time_since_recent_delinquency | avg_num_times_12m_delinquency | pct_no_12m_delinquency |
+---------------+-----------------------------------+-------------------------------+------------------------+
|       P1      |               9.9515              |             0.5287            |         84.16          |
|       P2      |              11.4225              |             0.3917            |         86.36          |
|       P3      |              13.3318              |             0.6343            |         76.52          |
|       P4      |              16.3047              |             0.7205            |         76.78          |
+---------------+-----------------------------------+-------------------------------+------------------------+

### **The Recency Factor Confirms the Pattern**

Perfect! This analysis confirms the key insight:

P1 borrowers show the LOWEST recent delinquency rates - only about 16% have had any delinquency in the last 12 months, and those who did had them a long time ago (10 months on average).

P4 borrowers, meanwhile, have higher recent delinquency rates and those delinquencies are more recent (average 16 months ago).

This solves the earlier paradox: P1 borrowers may have bad ancient history, but they've been clean recently. P4 borrowers have more recent problems. Lenders care more about what you did last year than what you did ten years ago.

---

### **Testing the "Gateway" Theory**

Some banks use "gateway" loans (small consumer loans) to introduce people to credit before giving them larger personal loans. Let me test if this strategy actually works by comparing people who started with consumer loans versus those who went straight to personal loans.

In [ ]:
%%sql
SELECT 
    CASE 
        WHEN Consumer_TL > 0 AND PL_TL = 0 THEN 'Consumer Only'
        WHEN Consumer_TL > 0 AND PL_TL > 0 THEN 'Both (Gateway User)'
        WHEN Consumer_TL = 0 AND PL_TL > 0 THEN 'PL Only (Skipped Gateway)'
        ELSE 'No Unsecured History'
    END AS borrower_segment,
    COUNT(*) as total_count,
    ROUND(SUM(CASE WHEN Approved_Flag = 'P1' THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_chance_of_P1,
    ROUND(SUM(CASE WHEN Approved_Flag = 'P4' THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_chance_of_P4
FROM final_credit_data
GROUP BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+---------------------------+-------------+------------------+------------------+
|      borrower_segment     | total_count | pct_chance_of_P1 | pct_chance_of_P4 |
+---------------------------+-------------+------------------+------------------+
| PL Only (Skipped Gateway) |     3603    |      23.18       |      11.52       |
|       Consumer Only       |    18448    |       7.16       |      15.19       |
|    No Unsecured History   |    24270    |      11.83       |       7.26       |
|    Both (Gateway User)    |     5015    |      15.47       |      17.99       |
+---------------------------+-------------+------------------+------------------+

### **The Gateway Strategy Might Need Rethinking**

The results here challenge conventional wisdom about credit building:

People who skipped the "gateway" and went straight to personal loans actually have the HIGHEST chance of being elite (P1) at 23%!

Meanwhile, people who only have consumer loans (the "gateway only" group) have the LOWEST chance of being P1 at just 7%.

What this suggests is that the highest-quality borrowers don't need training wheels - they can handle larger loans from the start.

The most interesting group is the "both" category - people who used the gateway AND got personal loans. They have decent P1 rates (15%) but also the highest P4 rates (18%). This suggests the gateway works for some but leads others into debt trouble.

This makes me question whether forcing everyone through small loans first is the best strategy. Maybe we should identify the naturally responsible borrowers and give them what they need from the start.

---

### **The Young Borrower Question**

Now, let me look at a specific segment: young borrowers (28 or younger) with new credit (oldest loan less than 2 years). This is the "new to credit" segment that many banks want to grow. Can we safely lend to them based on job stability?

In [ ]:
%%sql
SELECT 
    CASE WHEN Time_With_Curr_Empr >= 36 THEN 'Stable (>3yrs)' ELSE 'Unstable (<3yrs)' END AS job_stability,
    COUNT(*) AS total_prospects,
    ROUND(SUM(CASE WHEN Approved_Flag IN ('P1', 'P2') THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS pct_chance_of_approval
FROM final_credit_data
WHERE Age_Oldest_TL < 24 AND Age <= 28
  AND (Time_With_Curr_Empr / 12) <= Age 
GROUP BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

2 rows affected.

+------------------+-----------------+------------------------+
|  job_stability   | total_prospects | pct_chance_of_approval |
+------------------+-----------------+------------------------+
|  Stable (>3yrs)  |       7943      |         63.45          |
| Unstable (<3yrs) |       1772      |         59.65          |
+------------------+-----------------+------------------------+

### **Young Borrowers Need More Than Just Job Stability**

This analysis reveals an important limitation: job stability alone isn't enough to make young, new-to-credit borrowers safe.

Even young borrowers with 3+ years at their job only have a 63% chance of being in the safe categories. That's better than the 60% for those with less job stability, but it's still not great.

What this tells me is that for young borrowers, we need to look beyond traditional stability markers. Maybe we should consider:
- Their education level (though we saw that's not a great predictor either)
- Their spending patterns
- Their savings behavior
- Or we could start them with very small limits and watch their behavior

The key insight here is that we can't just take a "check the box" approach with young borrowers. They need more nuanced assessment, or we need to accept higher risk and price for it appropriately.

---

### **Testing Credit "Seasoning"**

Let me look at another factor: how long someone has had credit. In banking, this is called "seasoning" - the idea that a longer credit history gives more data points to assess risk. Does this hold true in our data?

In [7]:
%%sql
SELECT 
CASE WHEN Age_Oldest_TL < 12 OR Age_Oldest_TL IS NULL THEN '<1yr'
WHEN Age_Oldest_TL BETWEEN 12 AND 24 THEN '1-2yr'
WHEN Age_Oldest_TL BETWEEN 25 AND 36 THEN '2-3yr'
WHEN Age_Oldest_TL BETWEEN 37 AND 60 THEN '3-5yr'
WHEN Age_Oldest_TL BETWEEN 61 AND 120 THEN '5-10yr'
ELSE '>10yr' END AS oldest_tl_bucket,
AVG(Credit_Score) AS avg_credit_score
FROM final_credit_data
GROUP BY 1
ORDER BY FIELD(oldest_tl_bucket, '<1yr', '1-2yr', '2-3yr', '3-5yr', '5-10yr', '>10yr');


Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

6 rows affected.

+------------------+------------------+
| oldest_tl_bucket | avg_credit_score |
+------------------+------------------+
|       <1yr       |     669.6564     |
|      1-2yr       |     674.0413     |
|      2-3yr       |     675.4209     |
|      3-5yr       |     678.9036     |
|      5-10yr      |     690.0949     |
|      >10yr       |     709.8482     |
+------------------+------------------+

### **Time Builds Trust - But Slowly**

The data shows a clear relationship: the longer someone has had credit, the higher their credit score tends to be.

What's interesting is the pattern of growth: 
- In the first 5 years, scores increase slowly (from 670 to 679)
- After 5 years, there's a noticeable jump to 690
- After 10 years, scores average 710 - that's a 40-point advantage over someone with less than a year of history

This tells me that credit scoring algorithms really do reward longevity. But it's not just about age - it's about having a long track record of good behavior.

What's really valuable here is identifying the 5-year mark as a key milestone. Once someone has 5+ years of credit history without major issues, they enter a different risk category.

---

### **The "Finisher" vs. "Starter" Question**

Now let me look at an interesting behavioral pattern: what percentage of loans has someone successfully closed? I'm curious if people who finish what they start (close loans) have better credit than people who just keep opening new ones.

In [10]:
%%sql
SELECT 
CASE WHEN pct_closed_tl < 0.25 OR pct_closed_tl IS NULL THEN '<25%'
WHEN pct_closed_tl BETWEEN 0.25 AND 0.5 THEN '25-50%'
WHEN pct_closed_tl BETWEEN 0.5 AND 0.75 THEN '50-75%'
ELSE '>75%' END AS pct_closed_tl_bucket,
AVG(Credit_Score) AS avg_credit_score
FROM final_credit_data
GROUP BY 1
ORDER BY FIELD(pct_closed_tl_bucket, '<25%', '25-50%', '50-75%', '>75%');

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+----------------------+------------------+
| pct_closed_tl_bucket | avg_credit_score |
+----------------------+------------------+
|         <25%         |     674.4055     |
|        25-50%        |     680.9125     |
|        50-75%        |     688.3567     |
|         >75%         |     682.6091     |
+----------------------+------------------+

### **The Sweet Spot for Loan Closure**

This analysis revealed a fascinating pattern that looks like an inverted U:

People who close 50-75% of their loans have the highest average credit scores (688). But people who close MORE than 75% actually see their scores dip slightly (to 683).

Why would this happen?

Here's my theory: closing loans shows you're responsible and can finish what you start. But closing TOO MANY loans might mean you're exiting the credit market or have a "thin file" with not enough active credit to assess.

The people in the 50-75% range are probably the healthiest - they have enough active credit to show they're participating in the system, but they've closed enough to prove they're responsible.

This suggests there's a balance between being an active borrower and being a responsible one.

---

### **Job Stability as a Predictor**

Let me test another stability factor: how long someone has been at their current job. Does long-term employment translate to better credit risk?

In [ ]:
%%sql
WITH employment_duration AS (
SELECT Approved_Flag,
CASE WHEN Time_With_Curr_Empr < 12 THEN '<1yr'
WHEN Time_With_Curr_Empr BETWEEN 12 AND 24 THEN '1-2yr'
WHEN Time_With_Curr_Empr BETWEEN 25 AND 36 THEN '2-3yr'
WHEN Time_With_Curr_Empr BETWEEN 37 AND 60 THEN '3-5yr'
WHEN Time_With_Curr_Empr BETWEEN 61 AND 120 THEN '5-10yr'
ELSE '>10yr' END AS employment_duration_bucket
FROM final_credit_data
)

SELECT employment_duration_bucket, COUNT(*) AS total_prospects,
ROUND(SUM(CASE WHEN Approved_Flag IN ('P1', 'P2') THEN 1 ELSE 0 END)/ COUNT(*) * 100, 2) AS approval_rate
FROM employment_duration
GROUP BY 1
ORDER BY FIELD(employment_duration_bucket, '<1yr', '1-2yr', '2-3yr', '3-5yr', '5-10yr', '>10yr');


Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

6 rows affected.

+----------------------------+-----------------+---------------+
| employment_duration_bucket | total_prospects | approval_rate |
+----------------------------+-----------------+---------------+
|            <1yr            |        36       |     58.33     |
|           1-2yr            |       1580      |     66.84     |
|           2-3yr            |       3416      |     67.56     |
|           3-5yr            |       7609      |     68.31     |
|           5-10yr           |      19150      |     72.78     |
|           >10yr            |      19545      |     79.21     |
+----------------------------+-----------------+---------------+

### **The 5-Year Employment Milestone**

The pattern here is clear and strong: job stability matters, but there's a threshold effect.

People with less than 1 year at their job have only a 58% chance of being in the safe categories. That makes sense - they're in a probation period in more ways than one.

What's interesting is that between 1-5 years, approval rates are fairly flat (66-68%). It's like once you pass the first year, you're in a stable zone.

But the real jump happens after 5 years - approval rates jump to 73%, and after 10 years, they hit 79%!

This tells me that long-term employment (5+ years) is a strong signal of stability. These are people who have weathered economic cycles and organizational changes. They're less likely to experience sudden income shocks.

From a lending perspective, this suggests we could offer better terms to people with 5+ years at their job, as they represent lower risk.

---

### **Building a Simple Risk Scoring System**

At this point, I've identified several key risk signals. Now let me see if I can create a simple, easy-to-understand risk score that combines the strongest signals. I'm going to give 1 point each for:
1. Being in the top 30% of recent credit enquiries
2. Having credit card utilization above 70%
3. Having any delinquency in the last 12 months

Let's see how this simple 3-point system predicts risk categories.

In [52]:
%%sql
WITH risk_indicators AS (
SELECT Approved_Flag,
CASE WHEN NTILE(10) OVER(ORDER BY enq_L3m) >=8 THEN 1 ELSE 0 END AS high_enquiry,
CASE WHEN CC_utilization > 0.7 THEN 1 ELSE 0 END AS high_cc_utilization,
CASE WHEN num_deliq_12mts > 0 THEN 1 ELSE 0 END AS has_delinquency
FROM final_credit_data
),

aggregated_risks AS (
SELECT Approved_Flag, (high_enquiry + high_cc_utilization + has_delinquency) AS risk_score
FROM risk_indicators
)

SELECT risk_score,
ROUND(SUM(CASE WHEN Approved_Flag IN ('P1', 'P2') THEN 1 ELSE 0 END) / COUNT(*) * 100, 2) AS p1_p2_likelihood_pct,
ROUND(SUM(CASE WHEN Approved_Flag = 'P3' THEN 1 ELSE 0 END) / COUNT(*) * 100, 2) AS p3_likelihood_pct,
ROUND(SUM(CASE WHEN Approved_Flag = 'P4' THEN 1 ELSE 0 END) / COUNT(*) * 100, 2) AS p4_likelihood_pct
FROM aggregated_risks
GROUP BY 1
ORDER BY 1;


Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+------------+----------------------+-------------------+-------------------+
| risk_score | p1_p2_likelihood_pct | p3_likelihood_pct | p4_likelihood_pct |
+------------+----------------------+-------------------+-------------------+
|     0      |        88.46         |        8.91       |        2.63       |
|     1      |        56.10         |       22.05       |       21.86       |
|     2      |        42.96         |       24.26       |       32.78       |
|     3      |        47.91         |       22.83       |       29.26       |
+------------+----------------------+-------------------+-------------------+

### **The Power of Simple Signals**

This simple 3-point system is surprisingly powerful!

Here's what I found: borrowers with ZERO risk points have an 88% chance of being in the safe categories (P1 or P2). That's really high confidence from just three simple checks.

But here's where it gets interesting: with just ONE risk point, the probability of being in the high-risk P4 category jumps from 2.6% to 21.8%! That's nearly a tenfold increase in risk.

What this tells me is that these three signals are like early warning lights on a car dashboard. One warning light might be okay (maybe just a loose gas cap), but it definitely needs checking. Two or three warning lights mean you shouldn't drive that car until it's fixed.

The surprising thing is that 3-point borrowers actually have slightly better approval rates than 2-point borrowers. This might be because some high-income, high-spending people hit all three flags but have other compensating factors that keep them in P2.

---

### **The Business Impact Question**

Now I'm curious about the actual financial exposure. How much unsecured debt do these risk categories represent? And what are their incomes? This helps answer the "so what?" question for a bank's risk management team.

In [20]:
%%sql
WITH risk_indicators AS (
SELECT NETMONTHLYINCOME, Unsecured_TL,
CASE WHEN NTILE(10) OVER(ORDER BY enq_L3m) >=8 THEN 1 ELSE 0 END AS high_enquiry,
CASE WHEN CC_utilization > 0.7 THEN 1 ELSE 0 END AS high_cc_utilization,
CASE WHEN num_deliq_12mts > 0 THEN 1 ELSE 0 END AS has_delinquency
FROM final_credit_data
),

aggregated_risks AS (
SELECT NETMONTHLYINCOME, Unsecured_TL, (high_enquiry + high_cc_utilization + has_delinquency) AS risk_score
FROM risk_indicators
)

SELECT risk_score,
SUM(Unsecured_TL) AS total_unsecured_TL,
ROUND(AVG(NETMONTHLYINCOME), 2) AS avg_net_monthly_income
FROM aggregated_risks
GROUP BY 1
ORDER BY 1;

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/db_p1'

4 rows affected.

+------------+--------------------+------------------------+
| risk_score | total_unsecured_TL | avg_net_monthly_income |
+------------+--------------------+------------------------+
|     0      |       42673        |        25583.32        |
|     1      |       42259        |        27081.48        |
|     2      |       16163        |        29565.46        |
|     3      |        2280        |        32474.09        |
+------------+--------------------+------------------------+

### **The High-Income Risk Paradox**

This is one of the most important findings in the entire analysis:

The riskiest borrowers (those with 3 risk points) actually have the HIGHEST average income - about ₹32,474 per month! That's 27% higher than the safest group's average of ₹25,583.

What this means is that we can't use income as a safety blanket. Some of the highest earners are also the most financially reckless.

From a business perspective, the riskiest 3-point group represents 2,280 accounts, but because they have higher incomes, they probably have larger loan amounts. A default from this group would be much more damaging than a default from a low-income borrower.

Meanwhile, the 1-point group represents over 42,000 accounts - that's where most of the bank's exposure is. These are people who have triggered one warning sign but are still mostly in the safe categories. This is the group where good risk management can make the biggest difference.

---

###  **What I Learned: The Complete Credit Risk Story**

Looking back at this entire analysis, here's the story that emerged:

1. **Stability beats pedigree** - Marriage is a better predictor than education, and job tenure matters more than income level.

2. **Behavior matters most** - Recent credit enquiries, current credit card utilization, and recent payment history are the three most powerful predictors. A simple 3-point system using these can identify 88% of safe borrowers.

3. **The high-income paradox** - Some of the riskiest borrowers actually earn the most. Financial discipline isn't taught in college or guaranteed by a big paycheck.

4. **Time builds trust** - Credit history length and job tenure both show threshold effects at around 5 years. Before that, everyone's somewhat risky. After that, stability becomes clearer.

5. **One size doesn't fit all** - The "gateway loan" strategy works for some but fails for others. Young borrowers need more than just job stability.

###  The Business Recommendations

If I were presenting this to a bank's credit committee, I'd suggest:

1. **Weight marital status higher** in scoring models - it's a strong stability signal
2. **Create a simple 3-flag system** for quick triage: high enquiries + high utilization + recent delinquency = high risk
3. **Don't be blinded by high income** - check the behavioral signals first
4. **Offer better rates for 5+ year job tenure** - it's a proven stability marker
5. **Rethink mandatory gateway loans** - let some borrowers skip to what they need

The big takeaway? Credit risk assessment needs to look at both capacity (income, assets) AND character (spending habits, payment history). And character often matters more.

---